From 0aca75f62033c47ade4248439605e28f7052012f Mon Sep 17 00:00:00 2001
From: Seungha Yang <seungha@centricular.com>
Date: Tue, 15 Aug 2023 19:55:34 +0900
Subject: [PATCH 2/3] h265parser: Allow partially broken hvcC data

Ignores parsing error on the last nalu of the array if the nalu type
is not VPS/SPS/PPS

Fixes: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/2905
Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/5188>

The part for test is backported to 1.22.6
---
 gst-libs/gst/codecparsers/gsth265parser.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/gst-libs/gst/codecparsers/gsth265parser.c b/gst-libs/gst/codecparsers/gsth265parser.c
index 220ceb908..5f9cb6afe 100644
--- a/gst-libs/gst/codecparsers/gsth265parser.c
+++ b/gst-libs/gst/codecparsers/gsth265parser.c
@@ -4987,13 +4987,15 @@ gst_h265_parser_parse_decoder_config_record (GstH265Parser * parser,
   g_assert (gst_bit_reader_get_pos (&br) == 23 * 8);
   for (i = 0; i < num_of_arrays; i++) {
     GstH265DecoderConfigRecordNalUnitArray array;
+    guint8 nalu_type;
     GstH265NalUnit nalu;
     guint16 num_nalu, j;
     guint offset;
 
     READ_CONFIG_UINT8 (array.array_completeness, 1);
     SKIP_CONFIG_BITS (1);
-    READ_CONFIG_UINT8 (array.nal_unit_type, 6);
+    READ_CONFIG_UINT8 (nalu_type, 6);
+    array.nal_unit_type = nalu_type;
 
     READ_CONFIG_UINT16 (num_nalu, 16);
 
@@ -5005,6 +5007,15 @@ gst_h265_parser_parse_decoder_config_record (GstH265Parser * parser,
           2, &nalu);
       if (result != GST_H265_PARSER_OK) {
         g_array_unref (array.nalu);
+        /* Ignores parsing error if this is the last nalu and not an essential
+         * nalu for decoding */
+        if (i + 1 == num_of_arrays && j + 1 == num_nalu &&
+            nalu_type != GST_H265_NAL_VPS && nalu_type != GST_H265_NAL_SPS &&
+            nalu_type != GST_H265_NAL_PPS) {
+          GST_WARNING ("Couldn't parse the last nalu, type %d at array %d / %d",
+              nalu_type, i, j);
+          goto out;
+        }
         goto error;
       }
 
@@ -5021,6 +5032,7 @@ gst_h265_parser_parse_decoder_config_record (GstH265Parser * parser,
     }
   }
 
+out:
   *config = ret;
   return GST_H265_PARSER_OK;
 
-- 
2.34.1

